File: src/app/parser/m3uParser.ts
import type { ChannelEntity } from "../models/entities";

export function parseM3uPlaylist(playlistId: string, rawText: string): ChannelEntity[] {
  const lines = rawText.split(/\r?\n/);
  const parsed: ChannelEntity[] = [];
  let pendingName = "";
  let pendingLogo = "";
  let pendingGroup = "";

  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i].trim();
    if (!line) continue;

    if (line.startsWith("#EXTINF:")) {
      pendingName = line.includes(",") ? line.slice(line.lastIndexOf(",") + 1).trim() : `Channel ${parsed.length + 1}`;
      pendingLogo = line.match(/tvg-logo="([^"]+)"/i)?.[1] ?? "";
      pendingGroup = line.match(/group-title="([^"]+)"/i)?.[1] ?? "";
      continue;
    }

    if (!line.startsWith("#") && /^https?:\/\//i.test(line)) {
      parsed.push({
        id: `${playlistId}-${parsed.length + 1}`,
        playlist_id: playlistId,
        channel_name: pendingName || `Channel ${parsed.length + 1}`,
        logo_url: pendingLogo,
        stream_url: line,
        group_name: pendingGroup,
      });
      pendingName = "";
      pendingLogo = "";
      pendingGroup = "";
    }
  }

  return parsed;
}

File: android/app/src/main/java/com/app/streamplayer/parser/M3uParser.java
package com.app.streamplayer.parser;

import com.app.streamplayer.models.ChannelModel;

import java.util.ArrayList;
import java.util.List;

public class M3uParser {

    public static List<ChannelModel> parse(long playlistId, String rawText) {
        List<ChannelModel> channels = new ArrayList<>();
        String[] lines = rawText.split("\\r?\\n");
        String name = "";
        String logo = "";
        String group = "";

        for (String rawLine : lines) {
            String line = rawLine.trim();
            if (line.isEmpty()) {
                continue;
            }
            if (line.startsWith("#EXTINF:")) {
                int commaIndex = line.lastIndexOf(',');
                if (commaIndex >= 0 && commaIndex + 1 < line.length()) {
                    name = line.substring(commaIndex + 1).trim();
                } else {
                    name = "Channel";
                }
                logo = extractAttr(line, "tvg-logo");
                group = extractAttr(line, "group-title");
                continue;
            }
            if (!line.startsWith("#") && (line.startsWith("http://") || line.startsWith("https://"))) {
                channels.add(new ChannelModel(
                        0,
                        playlistId,
                        name.isEmpty() ? "Channel" : name,
                        logo,
                        line,
                        group,
                        0
                ));
                name = "";
                logo = "";
                group = "";
            }
        }

        return channels;
    }

    private static String extractAttr(String line, String key) {
        String token = key + "=\"";
        int start = line.indexOf(token);
        if (start < 0) return "";
        start += token.length();
        int end = line.indexOf('"', start);
        if (end < 0) return "";
        return line.substring(start, end);
    }
}